test/e2e/categories_test.go

```diff
@@ -0,0 +1,186 @@
+
+		By("PASSED!")
+	})
+	It("Create a cluster with additional categories", func() {
+		Expect(namespace).NotTo(BeNil())
+		flavor := "additional-categories"
+
+		By("Creating a workload cluster with additional categories set")
+		deployClusterAndWait(
+			deployClusterParams{
+				clusterName:           clusterName,
+				namespace:             namespace,
+				flavor:                flavor,
+				clusterctlConfigPath:  clusterctlConfigPath,
+				artifactFolder:        artifactFolder,
+				bootstrapClusterProxy: bootstrapClusterProxy,
+				e2eConfig:             *e2eConfig,
+			}, clusterResources)
+
+		By("Assign additional categories to the vms")
+		expectedCategoryKey := getExpectedClusterCategoryKey(clusterName)
+		additionalCategory := getAdditionalCategoryFromEnvironment()
+		Expect(additionalCategory.key).ToNot(BeEmpty())
+		Expect(additionalCategory.value).ToNot(BeEmpty())
+		expectedCategories := map[string]string{
+			expectedCategoryKey:    defaultClusterCategoryValue,
+			additionalCategory.key: additionalCategory.value,
+		}
+		verifyCategoriesNutanixMachines(clusterName, namespace.Name, expectedCategories)
+
+		By("PASSED!")
+	})
+
+	It("Create a cluster linked to non-existing categories (should fail)", func() {
+		flavor = "no-nmt"
+		Expect(namespace).NotTo(BeNil())
+
+		By("Creating Nutanix Machine Template with invalid categories")
+		invalidProjectNMT := createDefaultNMT(clusterName, namespace.Name)
+		invalidProjectNMT.Spec.Template.Spec.AdditionalCategories = []infrav1.NutanixCategoryIdentifier{
+			infrav1.NutanixCategoryIdentifier{
+				Key:   defaultNonExistingAdditionalCategoryKey,
+				Value: defaultNonExistingAdditionalCategoryValue,
+			},
+		}
+		createNutanixMachineTemplate(ctx, createNutanixMachineTemplateParams{
+			creator:                bootstrapClusterProxy.GetClient(),
+			nutanixMachineTemplate: invalidProjectNMT,
+		})
+		By("Creating a workload cluster")
+		deployCluster(
+			deployClusterParams{
+				clusterName:           clusterName,
+				namespace:             namespace,
+				flavor:                flavor,
+				clusterctlConfigPath:  clusterctlConfigPath,
+				artifactFolder:        artifactFolder,
+				bootstrapClusterProxy: bootstrapClusterProxy,
+				e2eConfig:             *e2eConfig,
+			}, clusterResources)
+
+		By("Checking machine status is 'Failed' and failure message is set")
+		verifyFailureMessageOnClusterMachines(ctx, verifyFailureMessageOnClusterMachinesParams{
+			clusterName:            clusterName,
+			namespace:              namespace,
+			expectedPhase:          "Failed",
+			expectedFailureMessage: "not found in category",
+			bootstrapClusterProxy:  bootstrapClusterProxy,
+		})
+		By("PASSED!")
+	})
+})
+
+func getExpectedClusterCategoryKey(clusterName string) string {
+	return fmt.Sprintf("%s-%s", defaultClusterCategoryKeyPrefix, clusterName)
+}
+
+func verifyCategoriesNutanixMachines(clusterName, namespace string, expectedCategories map[string]string) {
```

Comment:
why is this function part of this file as compared to test_helpers.go since we have verifyFailureMessageOnClusterMachines as part of that file. not unclear on whats the guideline here